### ISSUE TITLE
Adding torrents is awfully slow

### ISSUE THREAD

**Issue by NSQY** [transmission-remote-gtk/transmission-remote-gtk#147]
When adding a significant amount of torrents, I have noted that the rate that they are being added is frankly painfully slow. I can add >= 300 torrents via a web panel (https://github.com/ronggang/transmission-web-control) almost instantly whereas this client takes minutes to do the same.

**Comment by LaserEyess** [comment]
Can you compile with `-Dbuildtype=debug` and then run with `TRG_SHOW_OUTGOING=1 TRG_CURL_VERBOSE=1 trasnmission-remote-gtk`? From what I can tell, sending rpc requests should be dispatched in a thread pool, asynchronously. So, it should be possible for many requests to be sent in a short amount of time. I'm curious if there's some bottleneck or what I assume is not true.

**Comment by NSQY** [comment]
I presume this is what you're after.
```
* Found bundle for host 192.168.1.56: 0x7f335400c770 [serially]
* Can not multiplex, even if we wanted to!
* Re-using existing connection! (#0) with host 192.168.1.56
* Connected to 192.168.1.56 (192.168.1.56) port 9091 (#0)
* Server auth using Basic with user 'multimedia'
> POST /transmission/rpc HTTP/1.1
[...]
Mark bundle as not supporting multiuse
```

**Comment by LaserEyess** [comment]
hm, HTTP1.1 with no multiplexing, that would certainly make things slower. Thanks for the info, I'll take a look into it. Do you use a reverse proxy or is it a direct connection to the daemon?

**Comment by LaserEyess** [comment]
Did some digging into this, I think this could be a combination of a couple things

1. The daemon does not support HTTP/2 or any sort of multiplexing or advanced features, therefore that is one bottleneck we can never resolve
2. Right now requests are done async with a threadpool each with its own curl handle, and not sharing resources (or connections) between them. There is no multiplexing and googling around this may be a very inefficient way to use libcurl

There's a way to fix 2, but I'm not sure what would be better. There's a `curl_shared` interface and a `curl_multi` interface. Right now it seems like http/2 multiplexing is limited to the `curl_multi` interface, but that doesn't exactly work with threading the way the current trg code works. Using `curl_shared` would work with multiple threads, but it may not give the performance benefits. 


TL;DR, this is less simple to solve than it seems, simply changing some `curl_easy_setopt` stuff won't work.

**Comment by TingPing** [comment]
My personal choice would be to use libsoup. It integrates with the glib mainloop and has a nicer api. But that's a bit of work.

**Comment by LaserEyess** [comment]
I think ultimately getting rid of the threading and going for integration of the mainloop is defintely something to strive for.

There are two options I see:
 - Switch to the `curl_multi_socket` interface and attach that to the mainloop: https://everything.curl.dev/libcurl/drive/multi-socket. This would require a lot of refactoring in `trg-client.c`, as well as me learning how glib's mainloop works and curl's multi interface, since I'm only familiar with the easy one.
 - Give up on that and use libsoup, which would also be a lot of refactoring in `trg-client.c` but much more realizable and designed to work well with glib's mainloop. libsoup is a much heavier dependency, but if it can add significant performance benefits it may be worth it.

Personally, I want to try option 1 if only because it would let me learn something new in both glib and curl. However, none of this actually answers the question if it will make RPC faster. We could make a hyper-optimized HTTP/2 client with whatever backend we want and it means absolutely nothing if the other end is an HTTP1.1 daemon. I wonder what that web client is doing differently here if it's just talking directly to the daemon...

**Comment by TingPing** [comment]
> libsoup is a much heavier dependency

It is a much smaller library. We already depend on `glib`, this does pull in a `glib-networking` dependency but thats going to be anywhere you have `glib`.

**Comment by TingPing** [comment]
You can test the performance of a libsoup backed version here: https://github.com/TingPing/transmission-remote-gnome

**Comment by LaserEyess** [comment]
Hm, when I briefly looked at the docs I thought libsoup was way more than an HTTP client, but you're right it is not that big. I suppose there's no reason to use libcurl at all besides my selfish desire to learn more about the API, which I could do on another project, technically. Still I think the first thing to consider here is whether any of this is worth it? I suspect the daemon may be more of the bottleneck here, and libevent doesn't support HTTP2 so that won't change.

**Comment by TingPing** [comment]
HTTP2 has little value in this situation.

Over localhost, where latency is effectively 0 and throughput is insanely high. HTTP/1.1 can do everything. Connection re-use is the only factor which you say is working. The protocol overhead does exist but were probably talking bytes per message difference.

**Comment by NSQY** [comment]
I would just like to clarify, that this is not an issue exclusively with an obscene amount of torrent files. Even with a single file, or just a handful, the speed (or lack there of) is very noticeable. Testing now with 5 random files (small), it took about 20 seconds to add all of them and during this time other torrents were not being visibly updated on the UI.

The web panel, on the other hand, took about 3 seconds to process 9 odd files which were then shown in this client seamlessly. I'm not entirely sure why the web panel sometimes shreds through torrents (as described in my original issue), and other times takes a bit longer (maybe there is a limitation with transmission that I'm hitting? Afaik it is single threaded). The only limit to the web panel that I have found is funnily enough from the browser itself it seems, attempting to process > 300~ ish files causes the tab to crash.

> Do you use a reverse proxy or is it a direct connection to the daemon?

The only abstraction should be Docker.

**Comment by LaserEyess** [comment]
If this is a local instance of the daemon then yes latency does not matter here nor does connection re-use nor does HTTP/2. I'm thinking @TingPing is right here, the thread-pool is the issue most likely. It's quite overkill for how we use this program.

We could switch to the `curl_multi*` interface and hook that up with glib's mainloop and get async through that, but I am quite unsure how that works, and it looks like libsoup abstracts all of that away for us, and the async API integrates perfectly with glib's mainloop. That and libsoup, through glib-networking, has libproxy support, so we would be able to "drop" a dep (indirectly) by switching to it and get proxy support without ifdeffing.  Oh and having the developer for libsoup review the code is also a nice bonus as well. :)

I think this might not be something I do right now, but maybe in the next couple of weeks I'll try this.
